### DISCUSSION TITLE
Difficultés à implémenter la dilatation thermique dans le module @DSL Implicit

### DISCUSSION THREAD

**Discussion by ClementTOSI** [thelfer/tfel#389] [General]
Bonjour, 

J'ai quelques difficultés à implémenter la dilatation thermique dans le module @DSL Implicit. J'ai essayé de deux manières (cf. 👍 ) :

---------------------------------------------------------------------------------------
@DSL Implicit;
@Behaviour PerfectPlasticity;
@Author Clément TOSI;
@Date 18/04/2023;
@Description{};

@Epsilon 1.e-16;
//@Theta 1;

@Brick StandardElastoViscoPlasticity{
  stress_potential : "Hooke" {
    young_modulus : "max(((1.e5)-(59.*(T-273.15)))*1.e6,1.e9)",
    poisson_ratio : 0.3
 //   thermal_expansion : 0.0, //5.96e-6,   👎 
 //   thermal_expansion_reference_temperature : 293.15
  },
  inelastic_flow : "Plastic" {
    criterion : "Mises",
    isotropic_hardening : "Linear" {R0 : "1.e6*(1014.0-0.543*(T-273.15))/(1.0+exp(0.00803*(T-273.15-588.6)))"},
    maximum_equivalent_stress_factor : 1.5,
    equivalent_stress_check_maximum_iteration_factor: 0.4
  }
};

@ComputeThermalExpansion "/DilatationThermiq.mfront";   👎 
---------------------------------------------------------------------------------------

où le fichier DilatationThermiq.mfront est :

 ---------------------------------------------------------------------------------------
@Parser   MaterialLaw;
@Law      ThermalExpansionCoefficient;
@Author  Clément TOSI;
@Date     18/04/2023;

// output of law
@Output a ;

@Constant ReferenceTemperature =  293.15;

@Function{
  a = 5.96e-6;
} // end of Function
---------------------------------------------------------------------------------------

Avez vous des idées pour m'aider à implémenter la dilatation thermique ?
 Merci d'avance, Clément

**Comment by thelfer** [comment]
Bonjour @ClementTOSI ,

Ceci devrait fonctionner:

```cxx
@brick StandardElastoViscoPlasticity{
stress_potential : "Hooke" {
young_modulus : "max(((1.e5)-(59.(T-273.15)))1.e6,1.e9)",
poisson_ratio : 0.3
thermal_expansion : "ThermalExpansionCoefficient.mfront", ￼
// thermal_expansion_reference_temperature : 293.15
},
inelastic_flow : "Plastic" {
....
}
};
```

**Reply by ClementTOSI** [reply]
Bonjour @thelfer,

J'ai essayé avec la modif que tu as proposé. Après compilation, j'ai essayé de le tester (.mtest) pour vérifier la valeur de la déformation thermique. Voici le .mtest : 

----------------------------------------------------------------------------------
@Author Clement TOSI;
@Date 18/04/2023;
@Description{"Test d'une loi elasto-viscoplastique"};

@Behaviour<generic> './src/libBehaviour.so' 'PerfectPlasticity' ;
@ExternalStateVariable 'Temperature' 773.15;

@Real 'v0'                                         1.e-2;

@ImposedStrain<function> 'EXX'                   'v0*t';
@ImposedStrain<function> 'EYY'                   '-0.5*v0*t';
@ImposedStrain<function> 'EZZ'                   '-0.5*v0*t';
@ImposedStrain<function> 'EXY'                   '0.';
@ImposedStrain<function> 'EXZ'                   '0.';
@ImposedStrain<function> 'EYZ'                   '0.';

@Times {0.,'200' in 20000};
----------------------------------------------------------------------------------

Si j'ai bien compris, le fichier .res donne la déformation totale (E) et la déformation élastique (Ee). Pour retrouver la valeur de la déformation thermique je fais : Exx-Eexx et lors de la réponse élastique du matériau je devrai retrouver  la déformation thermique mais ce n'est pas le cas (je trouve 0 lors de la réponse élastique).

**Reply by thelfer** [reply]
`MTest` va calculer des déformations à partir de l'état initial supposé équilibré (sauf si vous utilisez les mots clés `@Strain`, `@Stress`).  Tous  les solveurs travaillent de manière incrémentale.

Ici vous partez à chaud et votre température est constante, il n'y a donc pas d'évolution de la dilatation thermique.

Essayer de partir à froid en partant de la température de référence:

```
@ExternalStateVariable {-0.1: 293.15, 0: 773.15}
@ImposedStrain 'EXX' 'v0* max(t, 0)';
@Times {-0.1, 0, ....};
```

De manière générale, c'est un bonne pratique que de systématiquement partir d'un état de référence bien identifié, en particulier en grandes transformations où la densité massique joue un rôle important (rôle masqué par MFront, mais essentiel dans les formulations de passage ente l'état de référence et l'état courant).

**Reply by ClementTOSI** [reply]
OK ! Merci beaucoup ! Ca fonctionne !
